File: 23/e01.ts
import { getContent } from "./utils";

const CREATE_FILES = false;
const DAY = __filename.split(".")[0];
const TEST = false;

const TEST_INPUT_2 = `two1nine
eightwothree
abcone2threexyz
xtwone3four
4nineeightseven2
zoneight234
7pqrstsixteen`;

const NUMBERS = [
  "one",
  "two",
  "three",
  "four",
  "five",
  "six",
  "seven",
  "eight",
  "nine",
];

const input = getContent(
  TEST ? DAY + "-test.txt" : DAY + "-input.txt",
  CREATE_FILES,
);

const isNumeric = (c: string) => {
  return !isNaN(parseInt(c, 10));
};

const part_1 = (data: string) => {
  const lines = data.split("\n");
  const result = lines
    .map((l) => {
      const chars = [...l];
      const first = chars.find((c) => isNumeric(c));
      const last = [...chars].reverse().find((c) => isNumeric(c));
      const num = Number(`${first}${last}`);
      return num;
    })
    .reduce((prev, cur) => prev + cur, 0);
  return result;
};

const part_2 = (data: string) => {
  const lines = data.split("\n");
  const result = lines
    .map((l) => {
      const numbersOnLine = [...l].reduce((prev, c, idx) => {
        if (isNumeric(c)) return [...prev, Number(c)];
        const strNum = NUMBERS.findIndex(
          (num) => l.slice(idx, idx + num.length) === num,
        );
        if (strNum > -1) {
          return [...prev, strNum + 1];
        }
        return prev;
      }, [] as string[]);
      return Number(
        `${numbersOnLine[0]}${numbersOnLine[numbersOnLine.length - 1]}`,
      );
    })
    .reduce((prev, num) => prev + num, 0);
  return result;
};

console.log();
console.log("1:", part_1(input));
console.log("2:", part_2(TEST ? TEST_INPUT_2 : input));

File: 23/e01.c
#include "utils.h"
#include <assert.h>
#include <stdio.h>
#include <stdlib.h>

#define NUM "01"
#define BUFSIZE 1028

#define NUMS_LEN 9
char *NUMS[] = { "one", "two", "three", "four", "five", "six", "seven", "eight", "nine"};

int isNumeric(char c) {
  return (c >= '1' && c <= '9');
}

int parseNums(int *nums, char *s, int parseStrNums) {
  int numCount = 0;
  int sLen = strlen(s);
  for (int i = 0; i < sLen; i++) {
    if(isNumeric(s[i])) {
      nums[numCount] = s[i] - '0';
      numCount++;
    } else if(parseStrNums) {
      for(int j = 0; j < NUMS_LEN; j++ ) {
        if(strstr(&s[i], NUMS[j]) == &s[i]) {
          nums[numCount] = j + 1;
          numCount++;
        }
      }
    }
  }
  return numCount;
}

int parseLines(char **lines, size_t lineCount, int parseStrNums) {
  int result = 0;

  for(int i = 0; i < lineCount; i++) {
    int *nums = malloc(BUFSIZE * sizeof(char));
    int numCount = parseNums(nums, lines[i], parseStrNums);
    if(numCount > 0) {
      int num = nums[0] * 10 + nums[numCount - 1];
      result += num;
    }
    free(nums);
  }

  return result;
}

int main(int argc, char *argv[]) {
  char *filename = argc > 1 ? argv[1] : "e" NUM "-input.txt";
  char **lines = NULL;
  int lineCount = readLinesFromFile(filename, &lines);

  int result1 = parseLines(lines, lineCount, 0);
  int result2 = parseLines(lines, lineCount, 1);
  freeLines(lines, lineCount);

  printf("PART 1: %d\n", result1);
  printf("PART 2: %d\n", result2);

  return 0;
}
